tools: Initial version of IPC4 tx message logger using eBPF

ujfalusi opened the pull request:
The bpftrace script will tap into the entry of sof_ipc4_log_header() which is always called but only prints if dynamic debugging is enabled. It's parameter list is not expected to be changed.

The bpftrace must be installed to be able to use the script
sudo pacman -S bpftrace
sudo emerge -a bpftrace
sudo apt install bpftrace
sudo dnf install bpftrace

To use the script:
sudo ./tools/sof-ipc4-msg-trace.bt
or
sudo bpftrace tools/sof-ipc4-msg-trace.bt

It will start logging the sent messages (including payloads) and received notification with exception of 0x1b060000 - trace update.

To stop the logging, just terminate the script with CTRL+C

marc-hb reviewed (commented):
I don't know BPF but this seems readable so LGTM. Some kernel developer should review.

Could you move this to a new subdirectory? The `tools/` directory (one of the common, meaningless names...) is getting crowded. How about `tools/BPF/ipc4-msg-trace.bt`? We already know it's SOF from the repo name.

marc-hb commented:
> To use the script: ...

Perfect for a `tools/BPF/README-ipc-trace.rst`

ujfalusi commented:
Changes since v1:
- script moved as `tools/BPF/ipc4-msg-trace.bt`
- Readme file added

ujfalusi commented:
Changes since v2:
- Update the README.md file for the eBPF directory

marc-hb reviewed (approved):
Someone else must review the actual BPF code

marc-hb commented:
"Big string" support submitted:
- https://github.com/bpftrace/bpftrace/pull/3228

marc-hb commented:
> "Big string" support submitted:

> https://github.com/bpftrace/bpftrace/pull/3228


Merged.

ujfalusi commented:
> "Big string" support submitted:
> 
>     * [Support big strings bpftrace/bpftrace#3228](https://github.com/bpftrace/bpftrace/pull/3228)

Thanks for the heads up.

I think the big string is still short for us (payload can be 4096 bytes) and the bigstring (if I read it right) tops at 1024.
I also prefer to have 'configurable' line length.
We certainly need to keep eye on bpftrace now to see what new features it is bringing for us to use.

marc-hb commented:
> Someone else must review the actual BPF code

Ping?

marc-hb commented:
SOFCI TEST

ujfalusi commented:
Changes since v3:
- Change the payload dump from bytes to 32bit words to align with kernel format and for easier handling
- Update the README to note that some distros only have old version of bpftrace.

ujfalusi commented:
Changes since v4:
- readability cleanup, no functional change
